apps/condo/domains/notification/tasks.js

```diff
@@ -1,118 +1,166 @@
+    // SMS has higher priority over EMAIL.
+    if (!isEmpty(phone) && !transports.includes(SMS_TRANSPORT)) transports.push(SMS_TRANSPORT)
+    if (!isEmpty(email) && !transports.includes(EMAIL_TRANSPORT)) transports.push(EMAIL_TRANSPORT)
 
-    if (!isEmpty(phone)) {
-        return SMS_TRANSPORT
-    }
-    if (!isEmpty(email)) {
-        return EMAIL_TRANSPORT
-    }
-    // TODO(pahaz): we should chose the best transport for the message.
-    //  We can chose transport depends on the message.type?
-    //  or use something like message.user.profile.preferredNotificationTransport if user want to get messages from TG
-    if (!isEmpty(user.email)) {
-        return EMAIL_TRANSPORT
-    }
-    return SMS_TRANSPORT
+    if (!isEmpty(user.email) && !transports.includes(EMAIL_TRANSPORT)) transports.push(EMAIL_TRANSPORT)
+
+    if (!isEmpty(transports)) return transports
+
+    // NOTE: none of requirements were met for message state, so we can't send anything anywhere actually.
+    throw new Error(`No appropriate transport found for notification id: ${id}` )
 }
 
-async function deliveryMessage (messageId) {
+async function deliverMessage (messageId) {
     const { keystone } = await getSchemaCtx('Message')
+    const message = await Message.getOne(keystone, { id: messageId })
 
-    const messages = await Message.getAll(keystone, { id: messageId })
-    if (messages.length !== 1) throw new Error('message id not found or found multiple results')
-
-    const message = messages[0]
     if (message.id !== messageId) throw new Error('get message by id has wrong result')
 
-    const transport = await _choseMessageTransport(message)
+    const transports = await _choseMessageTransport(message)
 
-    if (message.id !== messageId) throw new Error('get message by id wrong result')
-    if (message.status !== MESSAGE_SENDING_STATUS && message.status !== MESSAGE_RESENDING_STATUS) {
-        return `already-${message.status}`
-    }
+    if (message.status !== MESSAGE_SENDING_STATUS && message.status !== MESSAGE_RESENDING_STATUS) return `already-${message.status}`
 
     const baseAttrs = {
         // TODO(pahaz): it's better to use server side fingerprint?!
         dv: message.dv,
         sender: message.sender,
     }
+    const processingMeta = { dv: 1, transport: transports[0], step: 'init' }
 
-    const processingMeta = { dv: 1, transport, step: 'init' }
     await Message.update(keystone, message.id, {
         ...baseAttrs,
         status: MESSAGE_PROCESSING_STATUS,
         deliveredAt: null,
         processingMeta,
     })
 
-    try {
-        const adapter = TRANSPORTS[transport]
-        const messageContext = await adapter.prepareMessageToSend(message)
-        processingMeta.step = 'prepared'
-        processingMeta.messageContext = messageContext
+    const failedMeta = []
+
+    for (const transport of transports) {
+
+        if (message.type !== 'REGISTER_NEW_USER') console.log('deliverMessage:', { messageId, message, transport })
```

Comment:
WTF?

Reply:
It's an early draft PR, that was for debugging purposes. removed